### PR TITLE
4764 connect hn web to backend (r50) http external endpoint

### PR DESCRIPTION
**Pull request by daveb-hni**
Added connection for consuming an external endpoint using Spring WebClient. This includes:

- Adding dependency
- Adding connection properties
- Setting up the WebClient connection as a Spring Bean
- Sending to sample enrollment endpoint.

### CODE REVIEW

File: backend/src/main/java/ca/bc/gov/hlth/hnweb/service/EnrollmentService.java [hunk 1]
```diff
@@ -20,28 +25,70 @@
...
+	private String userName;
+
+	@Value("${R50.user.password}")
+	private String userPassword;
+
+	@Value("classpath:${R50.cert.file}")
+	private Resource certFile;
+
+	@Value("${R50.cert.password}")
+	private String certPassword;
```
Comment (tschia) [inline backend/src/main/java/ca/bc/gov/hlth/hnweb/service/EnrollmentService.java:49]: Unused variables

File: backend/src/main/java/ca/bc/gov/hlth/hnweb/service/EnrollmentService.java [hunk 2]
```diff
@@ -20,28 +25,70 @@
...
-		return parseR50v2ToAck("MSH|^~\\&|RAIPRSN-NM-SRCH|BC00002041|HNWeb|moh_hnclient_dev|20211013124847.746-0700||ACK|71902|D|2.4\r\n" + 
-				"MSA|AE|20191108082211|NHR529E^SEVERE SYSTEM ERROR\r\n" + 
-				"ERR|^^^NHR529E");
-	}
+		logger.debug("Response Status: {} ; Message:\n{}", response.getStatusCode(), response.getBody());
 		
+		return parseR50v2ToAck(response.getBody());
+	}
+
+	private ResponseEntity<String> postEnrollmentRequest(String path, String data, String transactionId) {
```
Comment (tschia) [inline backend/src/main/java/ca/bc/gov/hlth/hnweb/service/EnrollmentService.java:76]: path is unused

File: backend/src/main/java/ca/bc/gov/hlth/hnweb/service/EnrollmentService.java [hunk 3]
```diff
@@ -55,4 +102,21 @@ private Message parseR50v2ToAck(String v2) throws HL7Exception {
...
+	 * MSH|^~\\&|HNWeb|BC01000030|RAIENROL-EMP|BC00001013|20200529114230|10-TEST|R50^Z06|20200529114230|D|2.4||
+
+		versus the spec version
+
+	 * MSH|^~\\&|HNWeb|BC01000030|RAIENROL-EMP|BC00001013|20200529114230|10-ANother|R50^Z06|20200529114230|D|2.4
+	 * @param r50v2
+	 * @return
+	 */
+	private String formatMessage(String r50v2) {
+		String r50v2RequiredFormat = r50v2.replaceFirst("\r", "||\r");
```
Comment (tschia) [inline backend/src/main/java/ca/bc/gov/hlth/hnweb/service/EnrollmentService.java:118]: no need to declare a variable, can just return the value

File: backend/src/main/java/ca/bc/gov/hlth/hnweb/config/WebClientConfig.java [hunk 4]
```diff
@@ -0,0 +1,96 @@
...
+	@Value("classpath:${R50.cert.file}")
+	private Resource certFile;
+
+	@Value("${R50.cert.password}")
+	private String certPassword;
+
+	@Bean("enrollmentWebClient")
+    public WebClient enrollmentWebClient() {
+
+	    HttpClient httpClient= HttpClient.create().secure(t -> t.sslContext(getSSLContext()));
```
Comment (tschia) [inline backend/src/main/java/ca/bc/gov/hlth/hnweb/config/WebClientConfig.java:53]: getSSLContext() could return null, not sure what HttpClient will do if it is

Comment (daveb-hni) [reply backend/src/main/java/ca/bc/gov/hlth/hnweb/config/WebClientConfig.java]: Updated as per all comments and discussion on error handling and props files

File: backend/src/main/java/ca/bc/gov/hlth/hnweb/controller/EnrollmentController.java [hunk 5]
```diff
@@ -44,13 +45,9 @@ public EnrollSubscriberResponse enrollSubscriber(@Valid @RequestBody EnrollSubsc
...
 		
-		//Convert request to R50 message and get it as HL7 v2 String. (This message is used to send to external endpoint wrapped in JSON)
-		R50 r50 = convertEnrollSubscriberRequestToR50(enrollSubscriberRequest);
-		
-		//Send to R50 endpoint
-		Message message = enrollmentService.enrollSubscriber(r50);
-		
-		//Convert R50 endpoint response to HN Web response
+		R50 r50 = convertEnrollSubscriberRequestToR50(enrollSubscriberRequest);		
+		final String transactionId = UUID.randomUUID().toString();
```
Comment (weskubo-cgi) [inline backend/src/main/java/ca/bc/gov/hlth/hnweb/controller/EnrollmentController.java:49]: I don't tend to bother adding final to local variables as it clutters up the code. Effectively, I think it's only a hint to developers as the compiler will infer this anyways.

File: backend/src/main/java/ca/bc/gov/hlth/hnweb/exception/HNWebException.java [hunk 6]
```diff
@@ -0,0 +1,27 @@
+package ca.bc.gov.hlth.hnweb.exception;
```
Comment (weskubo-cgi) [inline backend/src/main/java/ca/bc/gov/hlth/hnweb/exception/HNWebException.java:1]: I see this exception being thrown but we don't have a handler for it. We may want to add an application wide exception handling in the future.

File: backend/src/main/java/ca/bc/gov/hlth/hnweb/service/EnrollmentService.java [hunk 7]
```diff
@@ -20,28 +23,55 @@
 
 	private static final Logger logger = org.slf4j.LoggerFactory.getLogger(EnrollmentService.class);
 
+	public static String TRANSACTION_ID = "TransactionID";
```
Comment (weskubo-cgi) [inline backend/src/main/java/ca/bc/gov/hlth/hnweb/service/EnrollmentService.java:26]: This should be final.

**Review by weskubo-cgi — CHANGES_REQUESTED** [review]
Couldn't comment directly but it seems current best practice is to use spinal-case (i.e. hyphenated) naming for REST paths. So instead of enrollsubscriber or enrollSubscriber it should be enroll-subscriber.

Also, I think we discussed it would be good to add a webclient based unit test if that's the approach we are taking.

File: backend/src/main/java/ca/bc/gov/hlth/hnweb/config/HnWebConfig.java [hunk 8]
```diff
@@ -23,4 +23,6 @@ public HapiContext hapiContext() {
 	public Parser parser() {
 		return hapiContext().getGenericParser();
 	}
+
 }
+
```
[the remark this answers is not in the document]
Comment (daveb-hni) [reply backend/src/main/java/ca/bc/gov/hlth/hnweb/config/HnWebConfig.java]: Removed

File: backend/src/main/java/ca/bc/gov/hlth/hnweb/config/WebClientConfig.java [hunk 9]
```diff
@@ -0,0 +1,102 @@
...
+import ca.bc.gov.hlth.hnweb.exception.HNWebException;
+import io.netty.handler.ssl.SslContext;
+import io.netty.handler.ssl.SslContextBuilder;
+import reactor.core.publisher.Mono;
+import reactor.netty.http.client.HttpClient;
+
+@Configuration
+public class WebClientConfig {
+
+	private final Logger logger = LoggerFactory.getLogger("WebClientConfig");
```
[the remark this answers is not in the document]
Comment (daveb-hni) [reply backend/src/main/java/ca/bc/gov/hlth/hnweb/config/WebClientConfig.java]: Done

File: backend/src/main/java/ca/bc/gov/hlth/hnweb/controller/EnrollmentController.java (hunk 5 above)
Comment (daveb-hni) [reply backend/src/main/java/ca/bc/gov/hlth/hnweb/controller/EnrollmentController.java]: Removed final

File: backend/src/main/java/ca/bc/gov/hlth/hnweb/exception/HNWebException.java (hunk 6 above)
Comment (daveb-hni) [reply backend/src/main/java/ca/bc/gov/hlth/hnweb/exception/HNWebException.java:1]: As discussed leaving as is for now, so start up fails.

File: backend/src/main/java/ca/bc/gov/hlth/hnweb/service/EnrollmentService.java (hunk 7 above)
Comment (daveb-hni) [reply backend/src/main/java/ca/bc/gov/hlth/hnweb/service/EnrollmentService.java]: Finalised it

**Review by weskubo-cgi — APPROVED** [review]

**Pull request merged**